lib/authlogic/session/base.rb

```diff
@@ -1825,21 +1857,28 @@ def params_credentials
         controller.params[params_key]
       end
 
+      def headers_credentials
+        controller.headers[headers_key]
+      end
+
       def params_enabled?
-        if !params_credentials || !klass.column_names.include?("single_access_token")
-          return false
-        end
-        if controller.responds_to_single_access_allowed?
-          return controller.single_access_allowed?
-        end
-        params_enabled_by_allowed_request_types?
+        params_credentials && single_access_token_enabled?
+      end
+
+      def headers_enabled?
+        headers_credentials && single_access_token_enabled?
       end
 
-      def params_enabled_by_allowed_request_types?
-        case single_access_allowed_request_types
-        when Array
-          single_access_allowed_request_types.include?(controller.request_content_type) ||
-            single_access_allowed_request_types.include?(:all)
+      def single_access_token_enabled?
+        return false unless klass.column_names.include?("single_access_token")
+        return controller.single_access_allowed? if controller.responds_to_single_access_allowed?
+
+        single_access_token_allowed_by_request_type?
+      end
+
+      def single_access_token_allowed_by_request_type?
+        if single_access_allowed_request_types.is_a?(Array)
+          (single_access_allowed_request_types & [controller.request_content_type, :all]).any?
```

Comment:
I like your use of intersection here. We could even use `Set.intersect?`.

```suggestion
          Set.new(single_access_allowed_request_types).intersect?(
            Set[controller.request_content_type, :all]
          )
```

WDYT? If you change this, don't forget to `require 'set'`.